master/pkg/model/task_container_defaults.go

```diff
@@ -60,8 +66,28 @@ func validatePortRange(portRange string) []error {
 	return errs
 }
 
+// UnmarshalJSON implements the json.Unmarshaler interface.
+// Setting defaults here is necessary over our usual "Define a default struct and unmarshal into it"
+// strategy because there are places (resource pool configs) where we need to know if the task
+// container defaults were set at all or if they were not; if they were set then that resource
+// pool's task container defaults are used instead of the toplevel master config's settings.  To
+// know if the user set them at the resource pool level, the resource pool has to have a nullable
+// pointer, which is not compatible with our usual strategy for defaults.
+func (c *TaskContainerDefaultsConfig) UnmarshalJSON(data []byte) error {
```

Comment:
Damn it.. I left a really long comment on this but I didn't "add it" and when I submitted the review it went away but I'll try again..

So to be clear: this is defaulted here because you need to distinguish between not provided resource pool task container defaults and provided ones, but you need to default the provided ones that are missing these fields still? Took me a bit to grok that a resource pool config with nil task containter default json would never unmarshal into it's task container defaults and thus stay nil.. am I even getting it now lol? i want to try to think about a better way to do this..

Reply:
> you need to distinguish between not provided resource pool task container defaults and provided ones, but you need to default the provided ones that are missing these fields still

Yes, exactly.  So the "set the defaults and unmarshal into them" strategy does not work at all.

> Took me a bit to grok that a resource pool config with nil task containter default json would never unmarshal into it's task container defaults and thus stay nil (and not have null pointer panics here)

Yes; the json unmarshalling code is what allocates fields before calling UnmarshalJSON whenever it sees keys in the json blob that match the tagname from `json:"TAGNAME"`.

> i want to try to think about a better way to do this

This thinking and work has already been done; it's in the experiment config rework.  That machinery has the following order:

1. Validate user input, exactly as it is provided
2. For valid inputs, unmarshal into structs.  Nil pointers may be present for fields the user did not provide.
3. Merge any partial configs you'd like to fill gaps into the user-provided config.
4. Call WithDefaults() on the object to fill nil pointers that have defaults
5. When reading data from the config object, use a generated getter whose return type automatically is a pointer only in the case that a field is optional but has no default (for stronger compile-time checks, basically).

So in this case, the master config should be parsed exactly as it was provided.  Defaults on all provided objects are set by WithDefaults().  In this case, the top-level task_container_defaults will always be filled, even if the user does not provide it, but the per-resource-pool task_container_defaults will stay as nil pointers.  Only if the user provided them will WithDefaults() set defaults on them.

This has the additional benefit that this whole mess of trying to pre-unmarshal the experiment config so you can figure out the right task_container_defaults in order to actually unmarshal the experiment config goes away, because in the new system: steps 1 and 2 happen, then you read the resource_pool setting, then you pick which task_container_default you want, then you continue with steps 3-5.

Reply:
yep, totally on the same page. I guess what I meant was I don't see a good solution that isn't a hell of a lotta work 😄.